src/core/batch.rkt

```diff
@@ -82,71 +82,173 @@
       (unmunge root)))
   exprs)
 
-(define (expand-taylor batch)
-  (define vars (batch-vars batch))
+(define (expand-taylor input-batch)
+  (define vars (batch-vars input-batch))
+  (define nodes (batch-nodes input-batch))
+  (define roots (batch-roots input-batch))
+
+  ; Hash to avoid duplications
   (define icache (reverse vars))
   (define exprhash
     (make-hash (for/list ([var vars]
                           [i (in-naturals)])
                  (cons var i))))
-  ; Counts
   (define exprc 0)
   (define varc (length vars))
 
-  ; Translates programs into an instruction sequence of operations
-  (define (munge prog)
-    (define node ; This compiles to the register machine
-      (match prog
-        [(list '- arg1 arg2) `(+ ,(munge arg1) ,(munge `(neg ,arg2)))]
-        [(list 'pow base 1/2) `(sqrt ,(munge base))]
-        [(list 'pow base 1/3) `(cbrt ,(munge base))]
-        [(list 'pow base 2/3) `(cbrt ,(munge `(* ,base ,base)))]
-        [(list 'pow base power)
-         #:when (exact-integer? power)
-         `(pow ,(munge base) ,(munge power))]
-        [(list 'pow base power) `(exp ,(munge `(* ,power (log ,base))))]
-        [(list 'tan args) `(/ ,(munge `(sin ,args)) ,(munge `(cos ,args)))]
-        [(list 'cosh args) `(* ,(munge 1/2) ,(munge `(+ (exp ,args) (/ 1 (exp ,args)))))]
-        [(list 'sinh args) `(* ,(munge 1/2) ,(munge `(+ (exp ,args) (neg (/ 1 (exp ,args))))))]
-        [(list 'tanh args)
-         `(/ ,(munge `(+ (exp ,args) (neg (/ 1 (exp ,args)))))
-             ,(munge `(+ (exp ,args) (/ 1 (exp ,args)))))]
-        [(list 'asinh args) `(log ,(munge `(+ ,args (sqrt (+ (* ,args ,args) 1)))))]
-        [(list 'acosh args) `(log ,(munge `(+ ,args (sqrt (+ (* ,args ,args) -1)))))]
-        [(list 'atanh args) `(* ,(munge 1/2) ,(munge `(log (/ (+ 1 ,args) (+ 1 (neg ,args))))))]
-        [(list op args ...) (cons op (map munge args))]
-        [(approx spec impl) (approx spec (munge impl))]
-        [_ prog]))
+  ; Mapping from nodes to nodes*
+  (define mappings
+    (make-hash (map (λ (n) (cons n n)) (build-list (batch-nodes-length input-batch) values))))
```

Comment:
Why do you need to initialize the hash at all? I think it's fine if it starts empty? Or maybe it starts with just the variables?

Reply:
Also does this need to be a hash? I think it is just a vector, with the same length as the input batch.

Reply:
Make sense, changed to a vector